packages/rpc-methods/src/restricted/manageState.ts

```diff
@@ -99,17 +118,108 @@ export type ManageStateArgs = {
 
 export const STORAGE_SIZE_LIMIT = 104857600; // In bytes (100MB)
 
+type GetEncryptionKeyArgs = {
+  snapId: string;
+  mnemonicPhrase: string;
+};
+
+/**
+ * Get a deterministic encryption key to use for encrypting and decrypting the
+ * state.
+ *
+ * This key should only be used for state encryption using `snap_manageState`.
+ * To get other encryption keys, a different salt can be used.
+ *
+ * @param args - The encryption key args.
+ * @param args.snapId - The ID of the snap to get the encryption key for.
+ * @param args.mnemonicPhrase - The mnemonic phrase to derive the encryption key
+ * from.
+ * @returns The state encryption key.
+ */
+async function getEncryptionKey({
+  mnemonicPhrase,
+  snapId,
+}: GetEncryptionKeyArgs) {
+  return await deriveEntropy({
+    mnemonicPhrase,
+    input: snapId,
+    salt: STATE_ENCRYPTION_SALT,
+    magic: STATE_ENCRYPTION_MAGIC_VALUE,
+  });
+}
+
+type EncryptStateArgs = GetEncryptionKeyArgs & {
+  state: Json;
+};
+
+/**
+ * Encrypt the state using a deterministic encryption algorithm, based on the
+ * snap ID and mnemonic phrase.
+ *
+ * @param args - The encryption args.
+ * @param args.state - The state to encrypt.
+ * @param args.snapId - The ID of the snap to get the encryption key for.
+ * @param args.mnemonicPhrase - The mnemonic phrase to derive the encryption key
+ * from.
+ * @returns The encrypted state.
+ */
+async function encryptState({ state, ...keyArgs }: EncryptStateArgs) {
+  const encryptionKey = await getEncryptionKey(keyArgs);
+  return await encrypt(encryptionKey, state);
+}
+
+type DecryptStateArgs = GetEncryptionKeyArgs & {
+  state: string;
+};
+
+/**
+ * Decrypt the state using a deterministic decryption algorithm, based on the
+ * snap ID and mnemonic phrase.
+ *
+ * @param args - The encryption args.
+ * @param args.state - The state to decrypt.
+ * @param args.snapId - The ID of the snap to get the encryption key for.
+ * @param args.mnemonicPhrase - The mnemonic phrase to derive the encryption key
+ * from.
+ * @returns The encrypted state.
+ */
+async function decryptState({ state, ...keyArgs }: DecryptStateArgs) {
+  try {
+    const encryptionKey = await getEncryptionKey(keyArgs);
+    const decryptedState = await decrypt(encryptionKey, state);
+
+    assert(isValidJson(decryptedState));
+
+    return decryptedState as Record<string, Json>;
+  } catch {
+    throw ethErrors.rpc.internal({
+      message: 'Failed to decrypt snap state, the state must be corrupted.',
```

Comment:
Can we test this?

Reply:
It's tested already:

https://github.com/MetaMask/snaps-monorepo/blob/29be72b3f51c6726579f3c88ae87af3e5f745272/packages/rpc-methods/src/restricted/manageState.test.ts#L222